llvm/lib/CAS/OnDiskGraphDB.cpp

```diff
@@ -1287,14 +1342,15 @@ OnDiskGraphDB::open(StringRef AbsPath, StringRef HashName,
                     .moveInto(DataPool))
     return std::move(E);
 
-  return std::unique_ptr<OnDiskGraphDB>(
-      new OnDiskGraphDB(AbsPath, std::move(*Index), std::move(*DataPool)));
+  return std::unique_ptr<OnDiskGraphDB>(new OnDiskGraphDB(
```

Comment:
> I think we have a trade off here for how to fault in upstream content. You can fault in one by one but consider this use case: I am loading a big object graph but I am only interested in one section so I just load section nodes, look at its content and decide not to load all the child nodes. With single node cloning, we will missing all the child nodes (the trie node exists, but they points to nothing), and it might be unexpected for a later compiler instance that gets a cache hit but realize it can't materialized the entire object file.
> 
> You can also clone the entire tree but it would be expected that accessing a parent node with millions of child is going to be really slow if it need to fault in everything.

I think we should add explicit API in `ObjectStore`/`ActionCache` for "make sure the entire tree is materialized for this cache key and fail if that is not the case. We need this anyway for getting nodes from a network CAS, the issue of "a sub-node is missing" is present for such a CAS. So this would be controlled at a higher level, while this lower-level fault-in function would be per-node.

Reply:
The direction I'm going with is that there is a higher-level abstraction that manages storage growth; `OnDiskGraphDB` is an implementation part of that but not where the storage management occurs.